snapfaas/src/fs/mod.rs

```diff
@@ -539,6 +588,49 @@ impl<S: BackingStore> FS<S> {
         })
     }
 
+    pub fn create_service(&self, dpriv: Component, invoking: Component, service_info: ServiceInfo) -> Service {
+        STAT.with(|stat| {
+            let mut uid: UID = rand::random();
+            let service_info_bytes = serde_json::ser::to_vec(&service_info).unwrap();
+            while !self.storage.add(&uid.to_be_bytes(), &service_info_bytes) {
+                uid = rand::random();
+                stat.borrow_mut().create_retry += 1;
+            }
+            Service {
+                privilege: dpriv,
+                invoking,
+                object_id: uid,
+            }
+        })
+    }
+
+    pub fn invoke_service(&self, service: &Service) -> Result<ServiceInfo, ServiceError> {
+        CURRENT_LABEL.with(|current_label| {
+            PRIVILEGE.with(|opriv| {
+                STAT.with(|stat| {
+                    // implicit endorsement
+                    utils::endorse_with(&*opriv.borrow());
```

Comment:
Call endorse_with_full and remove the endorse_with_full in the high-level `invoke_service`.